src/lib.rs

```diff
@@ -6,231 +6,108 @@ extern crate sha3;
+            match instr {
+                HASHER => {
+                    if let Some(h) = hiter.next() {
+                        stack.push(T::from_hash(h.to_vec()));
+                    } else {
+                        return Err(format!("Proof requires one more hash in HASHER"));
+                    }
                 }
-            }
-            LEAF(keylength) => {
-                if let Some(Leaf(key, value)) = kviter.next() {
-                    // If the value is empty, we have a NULL key and
-                    // therefore an EmptySlot should be returned.
-                    if value.len() != 0 {
-                        stack.push(Leaf(
-                            NibbleKey::from(key[key.len() - *keylength..].to_vec()),
-                            value.to_vec(),
-                        ));
+                LEAF(keylength) => {
+                    if let Some(leaf) = kviter.next() {
+                        // If the value is empty, we have a NULL key and
+                        // therefore an EmptySlot should be returned.
+                        match leaf.value() {
+                            None => stack.push(T::new_empty()),
+                            Some(_) => {
+                                if leaf.value_length().unwrap() == 0usize {
+                                    stack.push(T::new_empty())
+                                } else {
+                                    stack.push(leaf)
+                                }
+                            }
+                        }
                     } else {
-                        stack.push(EmptySlot);
+                        return Err(format!(
+                            "Proof requires one more (key,value) pair in LEAF({})",
+                            keylength
+                        ));
                     }
-                } else {
-                    return Err(format!(
-                        "Proof requires one more (key,value) pair in LEAF({})",
-                        keylength
-                    ));
                 }
-            }
-            BRANCH(digit) => {
-                if let Some(node) = stack.pop() {
-                    let mut children = vec![Node::EmptySlot; 16];
-                    children[*digit] = node;
-                    stack.push(Branch(children))
-                } else {
-                    return Err(format!(
-                        "Could not pop a value from the stack, that is required for a BRANCH({})",
-                        digit
-                    ));
+                BRANCH(digit) => {
+                    if let Some(ref node) = stack.pop() {
+                        let mut b = T::new_branch();
+                        b.set_ith_child(*digit, node);
+                        stack.push(b)
+                    } else {
+                        return Err(format!(
+                "Could not pop a value from the stack, that is required for a BRANCH({})",
```

Comment:
Hm, these lines don't seem to have proper indentation in github's diff explorer. But since CI passed `cargo fmt...` I assume it's alright.

Reply:
I agree it's weird, but that's the output of my `cargo fmt`.